include/deal.II/algorithms/general_data_storage.h

```diff
@@ -0,0 +1,380 @@
+   * @p name does in fact point to existing data, then this is overwritten.
+   */
+  template <typename Type>
+  void
+  add_or_overwrite_reference(Type &entry, const std::string &name);
+
+  /**
+   * Return a reference to the object with given name. If the object does
+   * not exist, then the input @p arguments will be used to construct an object
+   * of the given @p Type.
+   */
+  template <typename Type, typename... Args>
+  Type &
+  get_or_add_object_with_name(const std::string &name, Args &... arguments);
+
+  /**
+   * Return a reference to the object with given name. If the object does
+   * not exist, then the input @p arguments will be used to construct an object
+   * of the given @p Type.
+   */
+  template <typename Type, typename... Args>
+  Type &
+  get_or_add_object_with_name(const std::string &name, Args &&... arguments);
+
+  /**
+   * Return a reference to the object with given name. This function
+   * throws an exception if either an object with the given name is not
+   * stored in this class, or if the object with the given name is not of
+   * compatible type.
+   */
+  template <typename Type>
+  Type &
+  get_object_with_name(const std::string &name);
+
+  /**
+   * Return a constant reference to the object with the given name. This
+   * function throws an exception if either the name does not exist or if the
+   * object stored in the class with the given name cannot be converted to
+   * the given type.
+   */
+  template <typename Type>
+  const Type &
+  get_object_with_name(const std::string &name) const;
+
+  /**
+   * Find out if we store an object with given name.
+   */
+  bool
+  stores_object_with_name(const std::string &name) const;
+
+  //@}
+
+  /**
+   * An entry with this name does not exist in the internal boost::any map.
+   */
+  DeclException1(ExcNameNotFound,
+                 std::string,
+                 << "No entry with the name " << arg1 << " exists.");
+
+  /**
+   * An entry with this name does not exist in the internal boost::any map.
+   */
+  DeclException1(ExcNameHasBeenFound,
+                 std::string,
+                 << "An entry with the name " << arg1 << " already exists.");
+
+  /**
+   * The requested type and the stored type are different.
+   */
+  DeclException3(ExcTypeMismatch,
+                 std::string,
+                 const char *,
+                 const char *,
+                 << "The stored type for entry with name \"" << arg1 << "\" is "
+                 << arg2 << " but you requested type " << arg3 << ".");
+
+private:
+  /**
+   * Arbitrary user data, identified by a string.
+   */
+  std::map<std::string, boost::any> any_data;
+};
+
+
+/*----------------- Inline and template methods -----------------*/
+
+
+#ifndef DOXYGEN
+
+
+template <class Stream>
+void
+GeneralDataStorage::print_info(Stream &os)
+{
+  for (auto it : any_data)
+    {
+      os << it.first << '\t' << '\t'
+         << boost::core::demangle(it.second.type().name()) << std::endl;
```

Comment:
I think it would be useful to explain in the documentation what this function outputs